File: lection04/app_java/src/main/java/ru/hse/lection04/businesslayer/connectivity/ConnectivityProvider.java
package ru.hse.lection04.businesslayer.connectivity;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import ru.hse.lection04.objects.ConnectivityInfo;

@RequiresApi(api = Build.VERSION_CODES.N)
public class ConnectivityProvider extends AbstractConnectivityProvider {
    protected static final String TYPE_UNKNOWN = "UNKNOWN";

    private static final String[] TRANSPORT_NAMES = {
            "CELLULAR",
            "WIFI",
            "BLUETOOTH",
            "ETHERNET",
            "VPN",
            "WIFI_AWARE",
            "LOWPAN",
            "TEST"
    };

    protected final ConnectivityCallback mConnectivityCallback = new ConnectivityCallback();


    public ConnectivityProvider(Context context) {
        super(context);
    }


    @Override
    protected void setActivation(boolean value) {
        super.setActivation(value);

        if (value) {
            mConnectivityManager.registerDefaultNetworkCallback(mConnectivityCallback);
        } else {
            mConnectivityManager.unregisterNetworkCallback(mConnectivityCallback);
        }
    }


    protected void networkChanged(Network network) {
        final ConnectivityInfo info = extractInfo(network);

        for (IListener listener: mListeners) {
            listener.networkUpdated(this, info);
        }
    }

    protected ConnectivityInfo extractInfo(Network network) {
        final ConnectivityInfo info = new ConnectivityInfo();

        if (network != null) {
            final NetworkCapabilities capabilities = mConnectivityManager.getNetworkCapabilities(mConnectivityManager.getActiveNetwork());
            if (capabilities != null) {
                info.type = extractType(capabilities);
            }
        }

        return info;
    }

    protected String extractType(NetworkCapabilities capabilities) {
        if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_CELLULAR];
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_WIFI];
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_ETHERNET];
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_VPN];
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_BLUETOOTH];
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE)) {
                return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_WIFI_AWARE];
            }
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_LOWPAN)) {
                return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_LOWPAN];
            }
        }

        return TYPE_UNKNOWN;
    }


    protected class ConnectivityCallback extends ConnectivityManager.NetworkCallback {
        @Override
        public void onCapabilitiesChanged(@NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
            super.onCapabilitiesChanged(network, networkCapabilities);

            networkChanged(network);
        }

        @Override
        public void onLost(@NonNull Network network) {
            super.onLost(network);

            networkChanged(null);
        }
    }
}

File: lection04/app_kotlin/src/main/java/ru/hse/lection04/businesslayer/connectivity/ConnectivityProvider.kt
package ru.hse.lection04.businesslayer.connectivity

import android.content.Context
import android.net.ConnectivityManager.NetworkCallback
import android.net.Network
import android.net.NetworkCapabilities
import android.os.Build
import androidx.annotation.RequiresApi
import ru.hse.lection04.objects.ConnectivityInfo

/**
 * Имплементация, для поддержания работы логики от 24 версии API
 */
@RequiresApi(api = Build.VERSION_CODES.N)
class ConnectivityProvider(context: Context) : AbstractConnectivityProvider(context) {
    companion object {
        protected const val TYPE_UNKNOWN = "UNKNOWN"
        private val TRANSPORT_NAMES = arrayOf(
                "CELLULAR",
                "WIFI",
                "BLUETOOTH",
                "ETHERNET",
                "VPN",
                "WIFI_AWARE",
                "LOWPAN",
                "TEST"
        )
    }


    protected val mConnectivityCallback = ConnectivityCallback()


    override fun setActivation(value: Boolean) {
        super.setActivation(value)
        if (value) {
            mConnectivityManager.registerDefaultNetworkCallback(mConnectivityCallback)
        } else {
            mConnectivityManager.unregisterNetworkCallback(mConnectivityCallback)
        }
    }

    /**
     * Обработать изменение состоянии сети
     * @param network Полученные данные
     */
    protected fun networkChanged(network: Network?) {
        val info = extractInfo(network)
        for (listener in listeners) {
            listener?.networkUpdated(this, info)
        }
    }

    /**
     * Превращаем полученные данные в наш объект
     * @param network Полученные данные
     * @return наш объект с ифнормацией о соединении
     */
    protected fun extractInfo(network: Network?): ConnectivityInfo {
        val info = ConnectivityInfo()
        if (network != null) {
            val capabilities = mConnectivityManager.getNetworkCapabilities(mConnectivityManager.activeNetwork)
            if (capabilities != null) {
                info.type = extractType(capabilities)
            }
        }
        return info
    }

    protected fun extractType(capabilities: NetworkCapabilities): String {
        if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_CELLULAR]
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_WIFI]
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_ETHERNET]
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_VPN]
        } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH)) {
            return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_BLUETOOTH]
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE)) {
                return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_WIFI_AWARE]
            }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_LOWPAN)) {
                return TRANSPORT_NAMES[NetworkCapabilities.TRANSPORT_LOWPAN]
            }
        }
        return TYPE_UNKNOWN
    }

    /**
     * Подписчик на изменение состояния покдлючения, при помощи
     */
    protected inner class ConnectivityCallback : NetworkCallback() {
        override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
            super.onCapabilitiesChanged(network, networkCapabilities)
            networkChanged(network)
        }

        override fun onLost(network: Network) {
            super.onLost(network)
            networkChanged(null)
        }
    }
}